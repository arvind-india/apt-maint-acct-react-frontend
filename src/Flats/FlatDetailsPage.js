import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
//import Select from 'react-select';

import {
    Form,
    Button,
//    FormGroup,
    FormText,
    Input,
    Label,
} from 'reactstrap'

import { flatActions as actions, alertActions } from '../_actions'
//import { MODULES } from '../_constants'

let module = 'flats' // module name

class FlatDetailsPage extends React.Component {

  constructor(props) {
    super(props)
    const { dispatch, match, location } = props
    let model = location.state.model // model supplied from list page
//    let ops = model.operations?model.operations:''
    let initializeModel = {
      id: model.id,
      block_number: model.block_number,
      flat_number: model.flat_number
    }
    this.state = {
      model: initializeModel,           // model to edit
      submitted: false,
      adding: match.params.id === "0"
    }
    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    dispatch(alertActions.clear())  // clear alert messages from other pages
  }
  handleSubmit(event) {
    const { model } = this.state
    const { dispatch } = this.props

    event.preventDefault()
    this.setState({ submitted: true })

    if ( this.changedProps().length === 0 ) {
      dispatch(alertActions.error('No changes found...'))
    } else if(this.canBeSaved()){
      dispatch(actions.saveChanges(model))
    } else {
      dispatch(alertActions.error('Missing data'))
    }
  }

  canBeSaved() { // check for changes in mModel, if changes present, it can save
    const { model } = this.state
    if (!model.block_number) return false // no block_number selected, if so, cannot save changes
    if (!model.flat_number) return false // no flat_number entered, if so, cannot save changes
    return true // can save changes
  }
  changedProps() {
    const { model } = this.state
    const { location } = this.props
    let modelDB = location.state.model
    let props = []

    // check for changes in mModel props
    for(const prop in model) {
      if( prop === 'id') continue // exclude 'id' from comparision
      if(modelDB[prop] !== model[prop]) { // if data is changed wrt data in database
        props.push(prop)
      }
    }
    return props;
  }
  handleChange(event) {
    const { name, value } = event.target
    const { model } = this.state
    this.setState({
        model: {
          ...model,
          [name]: value
        }
    })
  }
  render() {
    const { alert } = this.props
    return (
      <div>
        <h2>Flat Details</h2>
        {alert.message && <div className={`alert ${alert.type}`}>{alert.message}</div>}
        {this.show()}
      </div>
    )
  }
  show(){
    const { adding } = this.state
    const { authzn } = this.props
    let title = adding?'Add':authzn.allowsEdit?'Edit':'View'
    return <Form onSubmit={this.handleSubmit} className="grid-form">
      <fieldset>
  			<legend>{title}</legend>
        <div data-row-span="2">
          {this.showBlockNumber()}
          {this.showFlatNumber()}
        </div>
      </fieldset>
      <br/>
      <Button type="submit" color="primary" hidden={!authzn.allowsEdit} title="Save changes">Save</Button>
      <Button color="link"><Link to="/flats" className="text-danger" title="Go to Flats">Cancel</Link></Button>
    </Form>
  }
  showBlockNumber() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Block Number</Label>
        <Input
          type="text"
          name="block_number"
          value={model.block_number}
          placeholder="<block number here>"
          title="Block Number"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.block_number
          && <FormText color="danger">Block Number is required</FormText>}
			</div>
  }
  showFlatNumber() {
    const { submitted, model } = this.state
    return <div data-field-span="1">
				<Label>Flat Number</Label>
        <Input
          type="text"
          name="flat_number"
          value={model.flat_number}
          placeholder="Description here"
          className="inputField"
          onChange={this.handleChange}
        />
        {submitted && !model.flat_number
          && <FormText color="danger">Flat Number is required</FormText>}
			</div>
  }
}

function mapStateToProps(state) {
  const { alert, authorizations } = state
//  const { user } = authentication
  const authzn = authorizations[module]
  return {
//    user,
    alert,
    authzn
  }
}

const connectedDetailsPage = connect(mapStateToProps)(FlatDetailsPage)
export { connectedDetailsPage as FlatDetailsPage }
