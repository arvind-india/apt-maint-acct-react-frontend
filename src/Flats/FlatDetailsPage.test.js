// FlatsPage.test.js0]

import React from 'react'
import { shallow } from 'enzyme'

import { FlatDetails } from './FlatDetailsPage'

const mockFlats = [
  {
    id: 1,
    block_number: '0',
    flat_number: 'g1'
  },
  {
    id: 2,
    block_number: '1',
    flat_number: 'g4'
  }
]

const mockProps = [
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    flats: {
      items: []
    }
  },
  {
    alert: {message: "testingAlert"},
    authzn: {
      allowsAdd: true,
      allowsEdit: true,
      allowsDelete: true,
      condition: ''
    },
    clearAlert: jest.fn(),
    saveChanges: jest.fn(),
    error: jest.fn(),
    flats: {
      items: []
    }
  }
]
const mockLocations = [
  {
    state: {
      model: mockFlats[0],
      submitted: false,
      adding: false
    }
  },
  {
    state: {
      model: mockFlats[1],
      submitted: false,
      adding: false
    }
  }
]

const mocks = [
  {
    blockNumber: {name: 'block_number', value: mockFlats[0].block_number},
    flatNumber: {name: 'flat_number', value: mockFlats[0].flat_number}
  },
  {
    blockNumber: {name: 'block_number', value: mockFlats[1].block_number},
    flatNumber: {name: 'flat_number', value: mockFlats[1].flat_number}
  }
]

const component = shallow(<FlatDetails {...mockProps[0]} location={mockLocations[0]}/>)

describe('FlatsPage params test', () => {
  it('should render FlatsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
  it('should have model state', () => {
    expect(component.state('model')).toBeDefined()
  })
  it('should have submitted state', () => {
    expect(component.state('submitted')).toBeDefined()
  })
  it('should have adding state', () => {
    expect(component.state('adding')).toBeDefined()
  })
})


describe('FlatDetailsPage events test 1', () => {
  it('should call clearAlert on submit and no crash', () => {
    component.find('#flatDetailsForm').simulate('submit', { preventDefault() {} })
    expect(mockProps[0].clearAlert.mock.calls.length).toEqual(1)
  })

  it('should be called with required states as arguments', () => {
    component.find('#blockNumber').simulate('change', {target: mocks[0].blockNumber})
    component.find('#flatNumber').simulate('change', {target: mocks[0].flatNumber}) // calling again to toggle from '' to 'R'
    component.find('#flatDetailsForm').simulate('submit', { preventDefault() {} })
    expect(mockProps[0].saveChanges.mock.calls[1][0]).toEqual(mockFlats[0])
  })

})


const component2 = shallow(<FlatDetails {...mockProps[1]} location={mockLocations[1]}/>)
describe('FlatDetailsPage events test 2', () => {
  it('should be called with required states as arguments', () => {
    component2.find('#blockNumber').simulate('change', {target: mocks[1].blockNumber})
    component2.find('#flatNumber').simulate('change', {target: mocks[1].flatNumber}) // calling again to toggle from '' to 'R'
    component2.find('#flatDetailsForm').simulate('submit', { preventDefault() {} })
    expect(mockProps[1].saveChanges.mock.calls[0][0]).toEqual(mockFlats[1])
  })
})
