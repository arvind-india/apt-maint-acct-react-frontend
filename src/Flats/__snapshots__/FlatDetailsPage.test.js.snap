// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h2>
        Flat Details
    </h2>
    <FlashMessage
        color={undefined}
        delay={3000}
        text="testingAlert"
    />
    <Form
        className="grid-form"
        id="flatDetailsForm"
        onSubmit={[Function]}
        tag="form"
    >
        <fieldset>
            <legend>
                Edit
            </legend>
            <div
                data-row-span="2"
            >
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Block Number
                    </Label>
                    <Input
                        className="inputField"
                        id="blockNumber"
                        name="blockNumber"
                        onChange={[Function]}
                        placeholder="<block number here>"
                        tag="p"
                        title="Block Number"
                        type="text"
                        value="0"
                    />
                </div>
                <div
                    data-field-span="1"
                >
                    <Label
                        tag="label"
                        widths={
                            Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                        }
                    >
                        Flat Number
                    </Label>
                    <Input
                        className="inputField"
                        id="flatNumber"
                        name="flatNumber"
                        onChange={[Function]}
                        placeholder="Description here"
                        tag="p"
                        type="text"
                        value="g1"
                    />
                </div>
            </div>
        </fieldset>
        <br />
        <Button
            color="primary"
            disabled={true}
            hidden={false}
            tag="button"
            title="No changes to save"
            type="submit"
        >
            Save
        </Button>
        <Button
            color="link"
            tag="button"
        >
            <Link
                className="text-danger"
                replace={false}
                title="Go to Flats"
                to="/flats"
            >
                Cancel
            </Link>
        </Button>
    </Form>
</div>,
]
`;
